File: backend/hospital-backend/src/main/java/com/hospital/prescription/controller/PrescriptionController.java
package com.hospital.prescription.controller;

import com.hospital.common.dto.ApiResponse;
import com.hospital.common.dto.PagedResponse;
import com.hospital.prescription.dto.*;
import com.hospital.prescription.service.PrescriptionService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

@RestController @RequestMapping("/prescriptions") @RequiredArgsConstructor
@Tag(name = "Prescriptions", description = "Prescription management")
public class PrescriptionController {

    private final PrescriptionService prescriptionService;

    @PostMapping
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR')")
    @Operation(summary = "Create Prescription")
    public ResponseEntity<ApiResponse<PrescriptionResponse>> createPrescription(
            @Valid @RequestBody PrescriptionCreateRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success("Prescription created", prescriptionService.createPrescription(request)));
    }

    @GetMapping("/{id}")
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR', 'PHARMACIST', 'PATIENT')")
    @Operation(summary = "Get Prescription")
    public ResponseEntity<ApiResponse<PrescriptionResponse>> getPrescription(@PathVariable Long id) {
        return ResponseEntity.ok(ApiResponse.success(prescriptionService.getPrescriptionById(id)));
    }

    @GetMapping("/appointment/{appointmentId}")
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR', 'PHARMACIST', 'PATIENT')")
    @Operation(summary = "Get Prescription by Appointment")
    public ResponseEntity<ApiResponse<PrescriptionResponse>> getPrescriptionByAppointment(@PathVariable Long appointmentId) {
        return ResponseEntity.ok(ApiResponse.success(prescriptionService.getPrescriptionByAppointmentId(appointmentId)));
    }

    @GetMapping("/patient/{patientId}")
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR', 'PHARMACIST', 'PATIENT')")
    @Operation(summary = "Get Patient Prescriptions")
    public ResponseEntity<ApiResponse<PagedResponse<PrescriptionResponse>>> getPatientPrescriptions(
            @PathVariable Long patientId, @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {
        return ResponseEntity.ok(ApiResponse.success(prescriptionService.getPatientPrescriptions(patientId, page, size)));
    }

    @GetMapping("/doctor/{doctorId}")
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR')")
    @Operation(summary = "Get Doctor Prescriptions")
    public ResponseEntity<ApiResponse<PagedResponse<PrescriptionResponse>>> getDoctorPrescriptions(
            @PathVariable Long doctorId, @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {
        return ResponseEntity.ok(ApiResponse.success(prescriptionService.getDoctorPrescriptions(doctorId, page, size)));
    }
}

File: backend/src/controllers/prescription.controller.ts
import { Request, Response, NextFunction } from 'express';
import * as prescriptionService from '../services/prescription.service';

export const createPrescription = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const result = await prescriptionService.createPrescription(req.body);
    res.status(201).json({
      success: true,
      message: 'Prescription created successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getPrescription = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id, 10);
    const result = await prescriptionService.getPrescriptionById(id);
    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getPrescriptionByAppointment = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const appointmentId = parseInt(req.params.appointmentId, 10);
    const result = await prescriptionService.getPrescriptionByAppointmentId(appointmentId);
    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getPatientPrescriptions = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const patientId = parseInt(req.params.patientId, 10);
    const page = parseInt(req.query.page as string || '0', 10);
    const size = parseInt(req.query.size as string || '20', 10);

    const result = await prescriptionService.getPatientPrescriptions(patientId, page, size);

    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getDoctorPrescriptions = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const doctorId = parseInt(req.params.doctorId, 10);
    const page = parseInt(req.query.page as string || '0', 10);
    const size = parseInt(req.query.size as string || '20', 10);

    const result = await prescriptionService.getDoctorPrescriptions(doctorId, page, size);

    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getAllPrescriptions = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const page = parseInt(req.query.page as string || '0', 10);
    const size = parseInt(req.query.size as string || '20', 10);

    const result = await prescriptionService.getAllPrescriptions(page, size);

    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const updatePrescription = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id, 10);
    const result = await prescriptionService.updatePrescription(id, req.body);
    res.status(200).json({
      success: true,
      message: 'Prescription updated successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

